Added LogEntry class to scheduling

michael-isaev opened the pull request:
The class creates a log entry for a single instruction executing in
simulator. It's needed for a future logging of simulated runs.

Commit:
Added LogEntry class to scheduling

The class creates a log entry for a single instruction executing in
simulator. It's needed for a future logging of simulated runs.

nicmcd commented on paragraph/scheduling/log_entry.h:
```diff
@@ -0,0 +1,57 @@
+
+ private:
+  // String identifier for instruction.
+  std::string name_;
```
All these members seem to reflect everything that is in the instruction and the instruction FSM. Why not just use a function that takes an InstructionFsm reference and returns a string for the log entry? This is a lot of extra work and extra memory consumption.

michael-isaev replied:
done

Commit:
Added logger class to scheduler

Logger adds information about all the executed instrucitons in a CSV
file in format processor_id, instruction_name, opcode, ready_time,
start_time, execution_time.

nicmcd commented on paragraph/scheduling/instruction_fsm.cc:
```diff
@@ -199,4 +202,16 @@ shim::StatusOr<Subroutine*> InstructionFsm::PickSubroutine() {
+std::string InstructionFsm::ToCsv(const std::string& delimeter) const {
+  CHECK(instruction_->GetGraph() != nullptr);
+  std::string log_entry_str = absl::StrCat(
+      instruction_->GetGraph()->GetProcessorId(), delimeter,
```
delimiter

michael-isaev replied:
fixed

nicmcd commented on paragraph/scheduling/instruction_fsm.cc:
```diff
@@ -199,4 +202,16 @@ shim::StatusOr<Subroutine*> InstructionFsm::PickSubroutine() {
+      instruction_->GetGraph()->GetProcessorId(), delimeter,
+      instruction_->GetName(), delimeter,
+      OpcodeToString(instruction_->GetOpcode()), delimeter,
+      absl::SixDigits(time_ready_), delimeter,
```
I can't find documentation on SixDigits. Will is properly handle the accuracy we need? For example, suppose and instruction started at time 55.0s and ended 10ns later.

michael-isaev replied:
Well, it looks like the code is the documentation, like [here](https://github.com/abseil/abseil-cpp/blob/master/absl/strings/str_cat.h#L43). Unfortunately it seems that it'll break for less that 6 decimal digits of precision, which is worse than float32 (7.225 digits). I wonder if there is a good library that does double -> string conversion. Or alternatively, protobuf is the answer instead of CSV

michael-isaev replied:
updated so it gives 12 decimal digits of precision - just enough for 1ps per second

nicmcd commented on paragraph/scheduling/logger.h:
```diff
@@ -0,0 +1,69 @@
+ public:
+  // Logger factory that creates a Logger object given a filename with default
+  // empty string. If non-empty file name is given, Logger opens file to append
+  // to it. If it fails, it return fail status.
```
returns

michael-isaev replied:
fixed

nicmcd commented on paragraph/scheduling/graph_scheduler.h:
```diff
@@ -88,6 +89,10 @@ class GraphScheduler {
   // API is used.
   double current_time_;
 
+  // Logger that logs to file all the instruction timings during the graph
```
Leave the file part of the implementation in the Logger class. Just state that all information about the instruction timing is passed to the Logger.

michael-isaev replied:
fixed

nicmcd commented on paragraph/scheduling/instruction_fsm.h:
```diff
@@ -85,6 +85,8 @@ class InstructionFsm {
   double GetTimeFinished();
   void SetTimeFinished(double current_time);
 
+  std::string ToCsv(const std::string& delimeter = ",") const;
```
Technically, "CSV" stands for comma separated values.

michael-isaev replied:
Right, that's why default delimiter is comma. Yet excel offers you to pick a delimiter symbol, that's why it is a parameter

nicmcd commented on paragraph/scheduling/logger.h:
```diff
@@ -0,0 +1,69 @@
+/* Copyright 2020 Google LLC
```
2021

michael-isaev replied:
fixed

nicmcd commented on paragraph/scheduling/logger.h:
```diff
@@ -0,0 +1,69 @@
+class Logger {
+ public:
+  // Logger factory that creates a Logger object given a filename with default
+  // empty string. If non-empty file name is given, Logger opens file to append
```
Why append? If I re-run a simulation, wouldn't I expect fresh data only?

michael-isaev replied:
fixed

nicmcd commented on paragraph/scheduling/logger.cc:
```diff
@@ -0,0 +1,105 @@
+
+absl::Status Logger::AppendToCsv(const InstructionFsm& instruction_fsm) {
+  RETURN_IF_ERROR(OpenFile());
+  std::cout << instruction_fsm.ToCsv() << std::endl;
```
I don't think you should have the instruction FSM handle the format. This logger class specifies the format in the "InitializeCsv" function and the ToCsv function would need to track that format. Neither could be adjusted independently. I think you should move the ToCsv functionality into the Logger class.

nicmcd commented on paragraph/scheduling/logger.h:
```diff
@@ -0,0 +1,69 @@
+  bool IsAvailable();
+
+  // Set filename for the log
+  absl::Status SetFilename(const std::string& filename);
```
Do you think it is useful to have this functionality? You are opening yourself up to corner cases here possible without any real reason to do so. Did you have a reason to want to change the log file path?

Along the same lines, I think you should just enforce that the ostream is always open. That way you never need to check.

michael-isaev replied:
I want to be able to start logging not from the start of simulation (think warm up). This way seemed like the easiest to get such functionality. I kinda tested against changing filepath, and test kinda worked. I'm not tremendously good with IO processing, but it seems that there a handful of places where filename can be changed (constructor and this function), and a handful of situations where ostream can get opened/closed. So I don't see it as a big issue.

nicmcd commented on paragraph/scheduling/logger.cc:
```diff
@@ -0,0 +1,105 @@
+
+absl::Status Logger::AppendToCsv(const InstructionFsm& instruction_fsm) {
+  RETURN_IF_ERROR(OpenFile());
+  std::cout << instruction_fsm.ToCsv() << std::endl;
```
Is this a leftover DEBUG print?

michael-isaev replied:
caught red handed

nicmcd commented on paragraph/scheduling/logger.h:
```diff
@@ -0,0 +1,69 @@
+  ~Logger() = default;
+
+  // Flushes log to file and closes file
+  void FlushToFile();
```
Why would a user need to call this?

michael-isaev replied:
It's part of Domino effect. I wanted to implement logger setting after scheduler is created. So that required covering a case when you had logger, but then decided to change the filename. But that required to close old file before opening a new one, or test would fail because not all information would be synchronized with both files. In the end of the day, I use it in test, so I want it to be public. In general, I don't think there is any harm in providing a way to speed up flushing io buffer to file.

nicmcd commented on paragraph/scheduling/graph_scheduler.h:
```diff
@@ -42,7 +43,7 @@ class GraphScheduler {
   // Part of Public API with simulators
   static shim::StatusOr<std::unique_ptr<GraphScheduler>> Create(
-      Graph* graph);
+      Graph* graph, const std::string& filename = "");
```
What if instead of giving the string to the scheduler, you give it the unique_ptr to a logger (better yet, make a SetLogger() method). Then you can drop the IsAvailable function call altogether and just check the logger for nullptr. All Logger objects would actually be logging.

In libdes (my parallel sim framework) I have a vector of logger pointers. All log events go to all logger objects. That way the user can create multiple loggers. For example, they could make a logger with filters that only look at send/recv, and another logger that looks at collectives. I think I defined a vector of raw pointers and implemented an AddLogger() method.

michael-isaev replied:
I was thinking exactly about that, but then thought that it's not simple enough, so I decided to make a single logger per scheduler, for now. In that case, unique_ptr looks like an okay choice to make sure the logger is unique :)
You can add issue/feature_request for multiple loggers with filters, and we can add them later. For now I exposed interface to set the logger from scheduler, but didn't expose logger itself as I didn't see need.

nicmcd commented on paragraph/scheduling/logger_test.cc:
```diff
@@ -0,0 +1,170 @@
+#include "paragraph/scheduling/graph_scheduler.h"
+#include "paragraph/shim/test_macros.h"
+
+std::string get_testfile_name(const std::string& _basename) {
```
Google doesn't like preceding underscores, but Nic does :)

michael-isaev replied:
fixed here and in graph_test.cc

Commit:
Fixed PR comments

Commit:
Added option to set logger in scheduler

nicmcd commented on paragraph/scheduling/logger.h:
```diff
@@ -0,0 +1,73 @@
+                                 const std::string& delimiter = ",");
+
+  // Write log to CSV file
+  absl::Status AppendToCsv(const InstructionFsm& instruction_fsm);
```
s/AppendToCsv/LogInstruction

michael-isaev replied:
done

Commit:
Trimmed unnecessary parts of logger

Commit:
Renamed AddToLog to LogInstruction

nicmcd commented on paragraph/scheduling/logger.cc:
```diff
@@ -35,35 +37,14 @@ shim::StatusOr<std::unique_ptr<Logger>> Logger::Create(
-
-bool Logger::IsAvailable() {
+Logger::~Logger() {
   if (ostream_.is_open()) {
```
Under what scenario could this be open?

nicmcd reviewed (approved):
Approve with single comment.

Commit:
Updated confusing ostream_ name to log_stream_

nicmcd commented on paragraph/scheduling/logger.cc:
```diff
@@ -56,26 +56,27 @@ absl::Status Logger::LogInstruction(
 Logger::Logger(const std::string& filename)
     : filename_(filename) {
-  std::fstream ostream_;
+  std::fstream log_stream_;
```
Why are you redeclaring this here? This "log_stream_" will not be the same as "this->log_stream_". The names are aliased and by the way this stream doesn't get used :)

Commit:
Killed zombie code in the logger constructor
